### ISSUE TITLE
[2104.08691] The Power of Scale for Parameter-Efficient Prompt Tuning

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#59]
## 論文リンク
https://arxiv.org/abs/2104.08691

## 公開日（yyyy/mm/dd）
2021/04/18

## 概要
後で書く。

**Comment by yoheikikuta** [comment]
T5 で prefix text を入れてタスクを指定することで全て text generation として解く、というのが面白い設定だと思っていたが、高い精度を出すにはモデルを downstream タスクでモデルをがっつり学習する必要があるというのは汎用モデルの限界かなというのが感想だった。

最近 Twitter とかで prefix text の部分のみをタスクに合わせて学習することで高い精度を発揮できるようになったというこの論文が流れてきたのを見て、どういう感じにやってるのか知りたいなと思って論文を読むことにした。

---
ちなみに paper reading でも過去に T5 論文を紹介していた https://github.com/yoheikikuta/paper-reading/issues/55

**Comment by yoheikikuta** [comment]
prefix text を入れるところをこの論文では prompt design と呼んでいる。
これはタスクを記述する短いテキストを文頭に入れることでどのタスクかを判断させる（モデル自体は同一のモデル）手法。この短いテキストの取り扱いはその後に続く実際の text と全く同じで、つまり同じように embedding して同じようにモデルに入力される、というものになっている。

この論文が指摘する prompt design の問題点は以下。

- エラーを起こしやすく、人間の介在が必要（適切に prefix text を選ぶ必要がある）
- input に含める例（これは GPT-3 の文脈）の数に精度が依存するが、これはモデルが扱える input length で限界がある
- これが一番重要だが、down stream タスクに対して tune したモデルと比べると精度は遠く及ばない（GPT-3 175B で SuperGLUE を解くと tune した T5-XXL よりも 17.5 points も低い。パラメタ数は 16 倍ほどだが。）

この辺りの、もっというと最後の、問題点を解決するために prefix text のところを改善したいというのがモチベーション。

**Comment by yoheikikuta** [comment]
自分が勉強不足なだけだが、この prefix text の部分を工夫するという先行研究は既にいくつかあるようだ。

prefix text としてどのような単語を選ぶのが良いのかを（単語なので離散的に）探すアルゴリズムを入れるとか、transformer の各 layer の前に prefix activation というのを入れて input とその部分だけ downstream タスクで学習するとか、in/out に subnetwork を入れてそこだけ学習するとか。

正直これらをちゃんと読んでないので正確には理解してないが、ベースとなる巨大なモデルを downstream タスクで学習するのは大変なのでそこは freeze させるけど、タスクに合わせて学習できる部分を導入してタスク依存性をそこで吸収させたいというモチベーションだろうと伺える。

この論文ではシンプルに、downstream タスクごとに k 個の token を準備して input の text の前にこれを付加し、その embedding だけは end-to-end で学習するという「soft prompt」という手法を提案している。

詳しいことは後で見ていくとして、この論文の貢献は以下だと主張している。

- 大きな言語モデルにおいて、提案する prompt tuning はモデル全体を tuning する場合と同等程度の性能を発揮することを示した
- さまざま構成方法を試し、性能がスケールすることで改善することを示した
- domain shift の問題においては prompt tuning がモデル全体の tuning よりも優れていることを示した
- prompt ensembling を提案し、その効果を実証した

**Comment by yoheikikuta** [comment]
prompt tuning とは何かの話に入る。

まず、問題設定としては T5 と同じで全てのタスクを text generation で解くものとする。

テキスト入力は {$` x_1, x_2, ..., x_n `$} という token 列で表現されているとする。これらは $` X_e `$ という embedding matrix でそれぞれ e 次元の特徴ベクトルに embed されるものとする。この embedding matrix は普通に言語モデルの一部として学習するもので、ここまでは特に新しいことはない。言語モデル全体は従来通り pre-training で学習され、この論文の手法では以降はずっと freeze されるものとなる。

次に $` k `$ 個の token {$` p_1, p_2, ..., p_k `$} という prompt を考える。これは downstream タスクごとに準備される。なのでもし考慮したい downstream task が 10 個あったら、合計で 10k 個の token がこの prompt 用に準備されることになる。ある一つの downstream タスクを考えれば、この k 個の token は学習可能な $` P_e `$ という先程とは別の embedding matrix (だけど射影した先の特徴量次元は同じ) で embed されるものとなる。downstream タスクにおいては {$` p_1, p_2, ..., p_k, x_1, ..., x_n `$} という token が入力となり、その入力は embedding matrix  $` X_e P_e `$ で特徴ベクトル化され、それがそのまま言語モデルへの入力になるというものになっている。

図でそれっぽく書いたのが以下。
Task Prompts の A,B,C のものが $` k `$ 個ある prompt token になっている。

<p align="center">
  <img src="https://i.imgur.com/zNYMXpQ.png">
</p>

prompt token をどのように準備するかという点に関しては具体的な実装が公開されてないので分からない面もあるが、論文から素直に読み取るなら、通常の text の vocabulary とは異なる id を付与していくという感じになるのだろう。例えば k=20 で downstream タスク 10 個を考えるとき、通常の語彙が 10000 だとすると、タスク 1 に関して 10001 - 10020 の ID としてタスク 1 用の prompt token が提供され、10021 - 10040 の ID がタスク 2 の token として提供され、という具合に新たに 200 個の ID が払い出されるという感じ。

**Comment by yoheikikuta** [comment]
以上、という感じなのだが prompt token に対する design の観点は二つある。

1 つ目は embedding をどのように初期化するという観点。
学習するんだから random initialization しといて学習すればいい、というのがナイーブな考え方としてはある。しかし、元々の text embedding で pre-training しているものがあるので、それをうまく使う方が筋が良さそうな気がする。embed したあとは元の言語モデルに食わせるので、その意味で元の vocabolary に近いような embedding を使うのは親和性が高いと期待できる。ということで学習してある vocabulary （に対応する embedding ）をサンプリングしてそれを初期値として使うという方法もよさそう。classification task に対しては出力クラス（のテキスト）の embedding で初期化するという方法もある。これが出力したい内容なので、input にこれを入れとくことによって出力を良い感じに制限しましょうという魂胆。

2 つ目は prompt token をどのくらいの数いれるかという観点。
(embedding 次元) x (prompt token の数) だけ新たな学習パラメタが必要になるので、prompt token の数は少なくしたい。これは実験で決めるが 100 くらいになる。

**Comment by yoheikikuta** [comment]
ベースとなる言語モデルは T5 を使うのだが、ここで T5 での事前学習をそのまま使うのは今の目的からするとちょっとずれていそうという話になる。

何を言ってるんだ？という感じになるので、そもそも T5 でどのように事前学習をしていたかを思い出す。
T5 は encoder-decoder モデルで学習を効率的にするために以下のような input - output を要請する。

input: Thank you for inviting me to the party last week.
masked: Thank you <X> me to the party <Y> week.
output: <X> me to the party <Y> <Z>

input は mask して、出力する token を減らして学習を効率化するために最初の sentinel token から最後の sentinel token までが出力する token になる。
これを考えると、モデルが受け取るのは必ずどこかが mask された input で真の入力された文章を受け取ることはないし、出力は必ず最初が <X> のような sentinel token になっている。このような「不自然な」状況で学習したものはあとから具体的なタスクで再学習したりすれば克服できるものではあるが、この論文では prompt 部分だけでそれをやる必要があり、そこの表現力は限られているので、もっとうまいことできるとよさそうという期待がある。

このことを考慮して、この論文では以下の 3 つの方法を試している。

- 1 つ目は span corruption で学習した従来の T5 を使って、prompt の学習のみでどれだけうまいこと上記の懸念が解決されるかを調べてみるという方法。
- 2 つ目は span corruption で学習した従来の T5 を使うが、downstream タスクで期待する出力の先頭に <X> という sentinel token を入れるようにするという方法（つまり正解の出力の先頭に sentinel token を手で付与する）。これは downstream タスクを作為的に pre-training の状況に合わせていくという小手先の方法。
- 3 つ目は span corruption で学習した後に、追加で prefix LM を使って学習するという方法。これは入力文と出力文を考えとき、入力文には全部 attention を貼るけど出力の方は因果を考えて attention を貼らないようにして学習するというものだった。これも T5 の論文で出てくるもの。自然な（corrupt されてない）文章を対象にして、一部を入力としてその続きを出力として学習する。これは pre-training の一部で downstream タスクと独立なので、一回やって frozen させておけるものである。これで自然な出力を出してくれるようになると期待。追加の pre-training は 10K - 100K ステップやってみて比較する。

**Comment by yoheikikuta** [comment]
実験。

frozen するモデルは T5.1.1 (最初の T5 から pre-training のデータから全ての supervised のデータを抜き、hyperparameter を最適化し、ReLU の代わりに GeGLU を使う) で、デフォルトの事前学習方法は LM adapted version で、デフォルトの prompt token 数は 100 とする。prompt token が（先行研究と比べても）多いような気がするが、これはこの部分だけが learnable なので多めに取っているという気持ちのようだ。

データは T5 のものと同じように SuperGLUE タスク（言語理解の 8 つのタスク）を使用。

prompt の学習は {batch size: 32, learning late: 0.3, steps: 30000} で、val データセットに対して early stopping して良いものを選択。optimizer は Adafactor で weigth decay 1e-5 で β_2 0.8 で parameter scaling は切ってるとのこと。

こういう学習のセットアップをテキストで説明して実装を公開しないとかもう論外な気がするんだけど...

文句はおいといて、prompt tuning がどれくらい優れているかという結果が以下。
縦軸が SuperGLUE のスコア（平均）で、横軸は model parameters だがこれは frozen する言語モデルのパラメタ数になっている。
model tuning はモデル全体を downstream タスクに対して再学習しているもので、prompt design は離散単語空間で prompt text を入れているもの。prompt tuning が提案手法。
すごい結果になっている。T5/GPT-3 の時点では model tuning と prompt design ほどの乖離があってやっぱ downstream タスクに対してモデルを再学習しないと最高レベルの精度は達成できないよねというものだったが、prompt tuning はモデルサイズが 10^10 パラメタとかならなんと model tuning と同等の性能になるという結果になっている。マジか。

<p align="center">
  <img src="https://i.imgur.com/sCkRkLb.png">
</p>

**Comment by yoheikikuta** [comment]
デフォルトの設定は上で書いたが、これらは実験で比較すべき選択肢が存在するので、それを ablation study で実施したのが以下。先程と同じで縦軸が SuperGLUE のスコア（平均）で、横軸は model parameters だがこれは frozen する言語モデルのパラメタ数になっている。

(a) が prompt 初期化の方法でこれは random uniform は弱く、既存 vocabulary の結果から sample するか class label のテキストを使うのが有効であるという結果。

(b) は prompt length で、単に長くすればいいということではなく、20~100 がよさそう。モデルサイズが大きければ single prompt token でもかなりの精度を発揮できるというのはなかなか驚き。長くなると悪くなるのがどういう理解なのか分からないけど、先行研究でも似たようなパターンがあるな！って言ってるのみ。入力長が限られてるのでいくらでも長くはできないのだけど、元のテキストが削れちゃうとかじゃない限り害をなすのはちょっとよく理解できないっすね。

(c) は pre-training medhod で、これはやはり prefix LM で追加学習した LM adaptation が有効であるという結果。モデルパラメタが 10^9 までは span corruption 全くもってダメ（chance rate くらい）なのに、$`10^{10}`$ のサイズになった途端にめちゃくちゃ高い性能を発揮するようになるの、端的に言って全く謎。これ本当なの？ `Intuitively, models pretrained to read and write sentinel tokens are hard to apply directly to tasks of reading and writing text without sentinels` とのことだが、サイズがでかくなったら急にできるようになることの直感的な説明には全くなってないよな。というか $` 10^8 `$ くらいのときはある程度当たるので、途中でやたら下がるのは完全に理解不能なんだよな。（さすがにちゃんと確認してるだろうけど）実験ミスってると考えたくなるような結果。
（ここでの実験は複数回やってるので prompt tuning プロセスの randomness ではなく）もともと使ってる checkpoint による影響かも、と言ってる。やらないけど何個か同じような checkpoint 作ればもう少し調べられるかも、でお茶を濁している。

(d) は LM adaptation の学習ステップ数を変えた場合。これは基本的に学習すればするほどよさそうという結果。もうちょっと多めに学習したときにどうなるかまで実験してもいい気がするけど。ここでも 10K のときにモデルパラメタ $` 10^9 `$ くらいで落ち込んでいる。これもやっぱ分からんな〜。

<p align="center">
  <img src="https://i.imgur.com/lbj57JZ.png">
</p>

**Comment by yoheikikuta** [comment]
他の手法との比較もいろいろ書いてある。詳しく読んでないけど結構近いことをしてる論文もありそう。この論文が主張するところによると、真に base model を frozen させて少ないパラメタで downstream タスクを高い精度で解くことができ、それをデカいモデルで実験しました、というのが違いということになる。アイデア自体はそこまで special なものではないということ。

**Comment by yoheikikuta** [comment]
prompt 部分で indirect にタスクの情報を学ばせるので、overfit を防ぎやすいのではないかという仮説のもと、domain shift での robusetness を測る実験もしている。

zero shot の枠組みで、source dataset で prompt tuning をして、同じタスクだけど違うドメインのデータでテストをする。
タスクとしては二つの質問文を入力として、それらが duplicates か否かを予測する、というものを対象にする。
一つのデータセットは Quora から収集した Quora Question Pairs (QQP) で、もう一つは news から収集した Microsoft Research Paraphrase Corpus (MRPC) を使用。

結果は以下で、より学習プロセスとしては重い model tuning と比べて accuracy の面で明確に良い結果を叩き出している。
domain shift のいろいろな手法のことを考慮しているわけではないが、単純な model tuning だと overfit してしまっているので、お手軽にそれを防げるということは言えそう。

<p align="center">
  <img src="https://i.imgur.com/awai0l8.png">
</p>

同様の実験を Machine Reading for Question Answering タスクでもやっていて、（雑に言って）同様の傾向を示していると言っている（model tuning に負けるやつもあるけど実はそれは domain として同じもの共有してるよね、とかもあったりする）。主張はまあ大体同じなので詳細は割愛。

**Comment by yoheikikuta** [comment]
もう一つ、prompt ensembling という手法も提案している。
これは単純に異なる prompt tuning を学習して、それを ensemble するというもの。ベースモデルは一つなので、でかいモデルを複数個準備することなく ensemble することが可能となる。

1 つの T5-XXL を frozen モデルとして、5 個の prompts を SuperGLUE タスクに対して学習し、ensemble として majority voting を使ってその性能を検証する。

結果が以下で、期待通りほぼ全てのタスクで single prompt の性能を上回った。
これは比較的結構お手軽にできるし精度を少しでも上げたいという場合には有効かもしれない。入力でごにょごにょやって ensemble で精度を上げるというと CLIP でやってたのを思い出すね。

<p align="center">
  <img src="https://i.imgur.com/FwVEe7n.png">
</p>

**Comment by yoheikikuta** [comment]
ということで一通り読み終わった。
soft (continuous) にして学習可能にせよ、という格言が prompt にも当てはまるという論文だった。
この方向性が発展すると、でかい pre-training モデルを省資源で有効活用しやすくるなるから小市民にはありがたいですね。
